pdm/models/specifiers.py

```diff
@@ -82,17 +82,17 @@ def __init__(self, version_str: str = "", analyze: bool = True) -> None:
         if version_str == "*":
             version_str = ""
         super().__init__(version_str)
-        self._lower_bound = self.MIN_VERSION  # type: Tuple[int, int, int]
-        self._upper_bound = self.MAX_VERSION  # type: Tuple[int, int, int]
-        self._excludes = []  # type: List[Tuple[Union[int, str], ...]]
+        self._lower_bound = self.MIN_VERSION
+        self._upper_bound = self.MAX_VERSION
+        self._excludes: List[Tuple[Union[int, str], ...]] = []
         if version_str and analyze:
             self._analyze_specifiers()
 
     def _analyze_specifiers(self) -> None:
         # XXX: Prerelease or postrelease specifiers will fail here, but I guess we can
         # just ignore them for now.
         lower_bound, upper_bound = self.MIN_VERSION, self.MAX_VERSION
-        excludes = set()  # type: Set[Tuple[Union[int, str], ...]]
+        excludes: Set[Tuple[Union[int, str], ...]] = set()
```

Comment:
might be nice to tidy this up with a type alias? what exactly is represented by a `Tuple[Union[int, str], ...]` here?

Reply:
Each item is a version tuple, which accepts the following forms:

- `(3, 8, 0)`
- `(3, 8, '*')`

Reply:
ah I see, so something along the lines of 

```python
VersionIncrement = Union[int, Literal['*']]
SemVer = Tuple[VersionIncrement, VersionIncrement, VersionIncrement]

...

excludes: Set[SemVer]
```

I'll have think about a tidy way to capture this.

almost worth a class.

Reply:
i feel like `pdm/models/specifiers.py` is crying out for some custom classes to encapsulate the version specifications. The code here looks challenging to debug/maintain

like this - https://docs.rs/semver/0.11.0/semver/struct.Version.html
and this - https://docs.rs/semver/0.11.0/semver/struct.VersionReq.html

I'm not planning to touch that in this pull request. will resolve this conversation for now

Reply:
Looks good. But be noted Literal is not supported in Python 3.7